File: java/src/com/nrkei/microservices/rapids_rivers/Packet.java
package com.nrkei.microservices.rapids_rivers;

/*
 * Copyright (c) 2016 by Fred George
 * May be used freely except for training; license required for training.
 * @author Fred George
 */

import com.google.gson.Gson;

import java.util.*;

// Understands a specific JSON-formatted message
public class Packet {

    final static String READ_COUNT = "system.read_count";

    private final Map<String, Object> jsonHash;
    private final Map<String, Object> recognizedKeys = new HashMap<>();

    Packet(Map<String, Object> jsonHash) {
        this.jsonHash = jsonHash;
        if (!jsonHash.containsKey(READ_COUNT)) jsonHash.put(READ_COUNT, -1.0);
        jsonHash.put(READ_COUNT, ((Double)jsonHash.get(READ_COUNT)).intValue() + 1);
    }

    void addAccessor(String requiredJsonKey) {
        if (!recognizedKeys.containsKey(requiredJsonKey))
            recognizedKeys.put(requiredJsonKey, jsonHash.get(requiredJsonKey));
    }

    public Object get(String key) {
        return recognizedKeys.get(key);
    }

    public void put(String key, Object value) {
        if (!recognizedKeys.containsKey(key))
            throw new IllegalArgumentException(
                    "Manipulated keys must be declared as required, forbidden, or interesting");
        recognizedKeys.put(key, value);
    }

    public String toJson() {
        Map<String, Object> updatedHash = new HashMap<>(jsonHash);
        for (String key : recognizedKeys.keySet())
            updatedHash.put(key, recognizedKeys.get(key));
        return new Gson().toJson(updatedHash);
    }

    public List getList(String solutionsKey) {
        return (List)get(solutionsKey);
    }
}

File: csharp_v2_beta/src/MicroServiceWorkshop/RapidsRivers/Packet.cs
/* 
 * Copyright (c) 2016 by Fred George
 * May be used freely except for training; license required for training.
 */

using System;
using System.Collections.Generic;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;

namespace MicroServiceWorkshop.RapidsRivers
{
    // Understands a well-formed message
    public class Packet
    {
        private readonly JObject _jsonHash;
        private readonly PacketProblems _problems;
        private readonly Dictionary<string, object> _recognized_keys = new Dictionary<string, object>();

        public Packet(string jsonString, PacketProblems problems)
        {
            _problems = problems;
            try
            {
                _jsonHash = JObject.Parse(jsonString);
            }
            catch (JsonException e)
            {
                _problems.SevereError("Invalid JSON format per NewtonSoft JSON library");
            }
            catch (Exception e)
            {
                _problems.SevereError("Unknown failure. Message is: " + e.Message);
            }
        }

        public void Require(params string[] requiredJsonKeys)
        {
            foreach (string key in requiredJsonKeys)
            {
                if (HasKey(key)) AddAccessor(key);
                else _problems.Error("Missing required key '" + key + "'");
            }
        }

        public void Forbid(params string[] forbiddenJsonKeys)
        {
            foreach (string key in forbiddenJsonKeys)
            {
                if (IsKeyEmpty(key))
                {
                    _problems.Information("Forbidden key '" + key + "' does not exist");
                }
                else _problems.Error("Forbidden key '" + key + "' actually exists");
            }
        }

        public bool HasKey(string key)
        {
            return _jsonHash[key] != null;
        }

        private bool IsKeyEmpty(string key)
        {
            JToken token = _jsonHash[key];
            return (token == null) ||
                (token.Type == JTokenType.Array && !token.HasValues) ||
                (token.Type == JTokenType.Object && !token.HasValues) ||
                (token.Type == JTokenType.String && token.ToString() == String.Empty) ||
                (token.Type == JTokenType.Null);
        }

        private void AddAccessor(string key)
        {
            if (_recognized_keys.ContainsKey(key)) return;
            switch (_jsonHash[key].Type)   // TODO: Where is polymorphism when you need it?
            {
                case JTokenType.String:
                    _recognized_keys[key] = (string)_jsonHash[key];
                    break;
                case JTokenType.Integer:
                    _recognized_keys[key] = (int)_jsonHash[key];
                    break;
                case JTokenType.Float:
                    _recognized_keys[key] = (float)_jsonHash[key];
                    break;
                default:
                    _recognized_keys[key] = _jsonHash[key];
                    break;
            }
        }

        public object Get(string key)
        {
            return _recognized_keys[key];
        }

        public void Put(string key, object value)
        {
            _recognized_keys[key] = value;
        }
    }
}